File: Code.java
/**
We have to watch out for two tree having same traversial but have different structure
For example:
    1
   /
  2
  
 and
 
    1
     \
      2
      
But we can counter it by placing a null value for example (2,x,1) and (x,2,1)
**/

boolean isSubtree(Tree<Integer> t1, Tree<Integer> t2) {
    if(t2 == null)
        return true;
    
    if(t1 == null)
        return false;
    
    StringBuilder st1 = new StringBuilder();
    StringBuilder st2 = new StringBuilder();
    
    makeString(t1, st1);
    makeString(t2, st2);
    
    System.out.println(st1 + " " + st2);
    
    return st1.indexOf(st2.toString()) != -1;
    
}

void makeString(Tree<Integer> t, StringBuilder sb){
    if(t == null)
    {
        sb.append(" x ");
        return;
    }
    
    sb.append(t.value + " ");
    makeString(t.left, sb);
    makeString(t.right, sb);
}

File: Code.c
//
//
// Definition for binary tree:
// typedef struct tree_##name {
//   type value;
//   struct tree_##name *left;
//   struct tree_##name *right;
// } tree_##name;
//
// tree_##name *alloc_tree_##name(type v) {
//   tree_##name *t = calloc(1, sizeof(tree_##name));
//   t->value = v;
//   return t;
// }

bool isEqual(tree_integer * t1, tree_integer * t2){
    return t1 && t2 ? 
        t1->value == t2->value && 
        isEqual(t1->left, t2->left) && 
        isEqual(t1->right, t2->right) 
        : t1 == t2;
}

bool isSubtree(tree_integer * t1, tree_integer * t2) {
    return t1 ? isEqual(t1, t2) || isSubtree(t1->left, t2) || isSubtree(t1->right, t2) : t2 == NULL;
}
